src/System.Private.CoreLib/shared/System/Buffers/ConfigurableArrayPool.cs

```diff
@@ -196,13 +195,16 @@ internal Bucket(int bufferLength, int numberOfBuffers, int poolId)
                 bool lockTaken = false, allocateBuffer = false;
                 try
                 {
+                    T[]?[] buffers = _buffers;
                     _lock.Enter(ref lockTaken);
 
-                    if (_index < buffers.Length)
+                    int index = _index;
+                    if ((uint)index < (uint)buffers.Length)
                     {
-                        buffer = buffers[_index];
-                        buffers[_index++] = null;
-                        allocateBuffer = buffer == null;
+                        buffer = buffers[index];
+                        buffers[index] = null;
+                        _index++;
```

Comment:
```suggestion
                        _index = index + 1;
```
as `index` is presumable in a register, so need to fetch from memory first.

But keep in mind @stephentoub's comment :wink:

Reply:
Tried that as well, similar answer as below